### ISSUE TITLE
stop bundling cairo

### ISSUE THREAD

**Issue by moodysalem** [software-mansion/scarb#420]
### Problem

i keep running into issues on the scarb version of cairo that are fixed on main, and it reminded me that in node, npm/node are not bundled together, and in foundry solc/foundry are not bundled together. i think you should just use the local cairo-test runner/cairo-run/starknet-compile/etc., if available, and implement some way to feed dependencies to it via the cli in the cairo repo

### Steps

for example, replace_class_hash is broken in cairo v2.0.0-rc2 but fixed on main, so i'm using a local cairo-test instead of scarb so my tests will pass

### Possible Solution(s)

stop bundling cairo if it's available on the path

### Version

```text
0.5.0-alpha
```

**Comment by mkaput** [comment]
Hello! This is a frequently reported issue that is actually an example of _what user wants is not necessarily what user needs_ scenario. Upfront, I declare that we understand the need, and we just have not yet got to the point of fulfilling it. But we are on track!

## Context

1. A compilation of Scarb workspace roughly consists of three abstract layers:
    - Scarb-only package resolution on the highest end
    - Cairo compiler-only compilation of the prepared `RootDatabase` on the lowest end
    - A _Scarb compiler_ in the middle, which functions as a bridge.
        - It prepares the aforementioned `RootDatabase`, collects diagnostics and builds output artefacts
        - There are actually two compilers in Scarb codebase currently, one for each target: `lib` and `starknet-contract`
        - The whole system is designed for users adding their own compilers, though! A user consumable extensibility system is just missing. Why? For example, to allow people to write their own Cairo-based blockchains.
    - Each layer _owns and takes responsibility of_ the lower one. This means that each Scarb compiler logically has its own version Cairo compiler inside. Of course, for sanity reasons, Scarb built-ins use the same Cairo compiler, but that's an important property coming from the extensibility feature.
    - Cargo/Foundry/NPM are different here! They do not have this middle layer.
    - Why do we have it, though? Because Cairo compiler is a single frontend that has many backends: one for pure Cairo and one for Starknet. And these backends ask for much higher level information than traditional compiler. Scarb targets/compilers mechanism is a way to expose this to the users.
2. Cairo compiler is also unique in a different way: *it is primarily a Rust crate*.
    - The binaries are just one of the possible _user_ interfaces, that is provided by compiler team. 
    - Scarb is another _user_ interface. 
    - What is the _machine_ interface, then? The `cairo-lang-*` crates are. And this is the only interface Scarb can reliably consume.

## Technical difficulties of implementing what user _wants_

So firstly, we have always been responding that Scarb and Cairo are welded together, and it's not possible to make it plug&play here. There is actually a theoretical solution, but let me maybe describe why detaching from Cairo compiler is not possible implementation-wise, or in other words - not feasible to a point of being practically impossible to do nicely and correctly:

1. First, the Cairo compiler APIs are not super well polished. For Scarb purposes, we regularly submit patches upstream which improve these APIs, enabling us to implement Scarb features.
    - This means, that we would have to backport all these patches to all past Cairo versions.
    - This is actually not possible in all cases, and would be painful in others.
2. Then, Scarb would have to maintain binary wrappers over Cairo compilers, again compatible with all Cairo versions existing, and separate binary for each Scarb target. This is probably tedious, though manageable.
3. We would have to think how to deal with cairo-test, Protostar, cairo-run, the language server and others. All these tools actually make a heavy use of Cairo crates.
4. And finally, some Scarb features only make sense with particular ranges of Cairo compiler versions, because things are added, deprecated or removed altogether. For example, [`casm-add-pythonic-hints`](https://docs.swmansion.com/scarb/docs/starknet/contract-target#casm-contract-class-generation) wasn't a thing in the earliest Cairo releases. Or even more: the [`starknet`](https://docs.swmansion.com/scarb/docs/starknet/starknet-package) package is planned to work completely different in the future than it is now.

I hope it should be clear at this point (if it is not, then this is my mistake): we have a multidimensional compatibility matrix, and we would have to develop, test and maintain forks/backports/binary wrappers for each cartesian product of this matrix. This is the infeasibility of this approach.

## What does the user _need_?

Two things:
1. > i keep running into issues on the scarb version of cairo that are fixed on main
2. an ability to compile code written in older Cairo, `1.0.0` especially

This is actually what users need. We haven't heard of any other reasons from the community, nor we cannot think of ourselves.

## How can we give users what they _need_?

Our team together with @amanusk, @spapinistarkware and the rest of StarkWare team are discussing the details how to tackle this issue, but I think I can already outline on high level what we aim to. **So, a notice: we may end up doing a completely different thing here.**

In short:
1. #271: Regular releases of Scarb that will come with a head of Cairo `main` branch.
2. #384: An ability to install multiple versions of Scarb and to conditionally run one matching the project

This is actually how Rust and Cargo works, you may be familiar with `rustup` and its components etc. 😄 

We also know that this solution has a shortcoming: it's not _backporting_ newer Scarb features back to older Cairo compilers. It doesn't matter for Rust, but actually matters for us, because Starknet promises 6 month compatibility period with Cairo releases. This seems to be a case-by-case situation, though, and we will treat it as such. For example, we plan to release Scarb 0.4.1 with cairo-run bundled (#430).

**Comment by mkaput** [comment]
Closing the issue because aforementioned issues technically supersede it, but feel free to reopen it 😃  I'm open for hot discussions.

**Comment by moodysalem** [comment]
I don't really get why it's so infeasible. I think you might think the user wants way more than they do.

> This means, that we would have to backport all these patches to all past Cairo versions.

Don't do this. Scarb only has one version of Cairo currently, so if a single scarb version only works with a handful of compatible Cairo versions that share an API/have the required features, it would still be a major improvement. Number 2 also does not apply if you assume a scarb version only works with some range of compatible cairo versions. The CLI interface should not change that often, and I'm assuming here Scarb calls cairo via the CLI.

Not sure why you would care about protostar, but you can just implement logic for downloading a particular cairo release's binaries for the appropriate platform and store the binaries somewhere local to scarb and use that to execute the test commands. In foundry, you specify the solidity version and it fetches it for you, but presumably it then just executes it via the CLI.

Cairo is breaking compatibility all the time. I don't care if the latest Scarb only supports the latest Cairo, given the older Cairo versions will eventually not be supported by the network at all, and I can just use an older version of Scarb if I really care. But I assume the CLI API will stabilize and the range of versions that will be supported by Scarb will increase after that.

**Comment by mkaput** [comment]
I think I have wrongly put an emphasis on the infeasibility of doing what you described, instead of straight away talking about the product side of things. I am sorry for that. 

Scarb **is a part of Cairo core DX**. This is fundamentally different from foundry/sol situation. Think of how Cargo is related to `rustc`. You install Rust 1.70 and get both things.  We aim to do similarly here. Inspired by your responses, I've actually accelerated conversations about synchronising Scarb and Cairo versions, to make this obvious, because I realised the current situation might indeed be suboptimal 🙂 

So, to give an example. If you want to use a different Cairo version that the one that comes with Scarb, we expect you to do a little extra work to fork Scarb and make your custom build. We still want to simplify this for general usage pattern, which is using latest `main` and nightlies will solve that.

Note: I know that `rustc` is a binary, and technically you could swap it in the way you ask for, but am not aware of anybody doing this. It's just that rustc is a lower level part of compilation stack, and here in Cairo, compiler APIs fulfill this role.

---

> The CLI interface should not change that often, and I'm assuming here Scarb calls cairo via the CLI.

> But I assume the CLI API will stabilize and the range of versions that will be supported by Scarb will increase after that.

These two assumptions do not hold.

Starkware is building these CLIs primarily for their internal usage, and we (Software Mansion & Starkware) always meant Scarb to be the user-facing interface. Starkware reserves the right to change these CLIs however they want in future releases.

**Comment by moodysalem** [comment]
> So, to give an example. If you want to use a different Cairo version that the one that comes with Scarb, we expect you to do a little extra work to fork Scarb and make your custom build. We still want to simplify this for general usage pattern, which is using latest `main` and nightlies will solve that.

How is this good for the Scarb product? Nobody wants to fork and rebuild npm to use a new version of node or vice versa. They have their own release schedules, different sets of bugs, different concerns, etc.

I don't know of any other package manager that actually bundles a specific version of the compiler/runtime. You install node and get NPM too, but you can upgrade either one separately, or use a different package manager like yarn or pnpm. As you said, Cargo comes with rust but you _can_ swap it out. It's not usual that you have to swap out node or rustc, but there's also no technical reason to tightly couple the two. Not sure why Scarb wants to reinvent the DX here.

> Starkware is building these CLIs primarily for their internal usage, and we (Software Mansion & Starkware) always meant Scarb to be the user-facing interface. Starkware reserves the right to change these CLIs however they want in future releases.

Adding arguments for configuring dependency resolution, or classpath, or whatever it's called in that language, is pretty normal for a compiler. I don't expect a compiler, or in general any binary, to change these arguments (the API) with every release.

**Comment by maciektr** [comment]
> Not sure why Scarb wants to reinvent the DX here.

IMHO shuffling underlying binaries is exactly opposite to the DX Cargo offers.

**Comment by moodysalem** [comment]
Between v0.5.2 being incompatible with testnet/mainnet and wanting to update my contracts for cairo v2.1.0 (scarb v0.6.0), I am changing scarb versions so often, like 5 times today already, and it's a huge pain because there are 4 binaries that are not code-signed. Meaning every time I download a new version I have to try to run them, click through a dialog, go to Apple privacy settings, allow the binary, try to run it again, and click open in the dialog. 

> IMHO shuffling underlying binaries is exactly opposite to the DX Cargo offers.

Shuffling binaries is exactly the DX scarb offers because 1 scarb version == 1 cairo version.

**Comment by spapinistarkware** [comment]
I tried asdf and I really enjoyed the experience!

**Comment by moodysalem** [comment]
> Hi @moodysalem! I know doing this manually is troublesome, fortunately **we do have** an _equivalent of a nvm for managing scarb versions_. Please take a look at our `asdf` plugin: https://docs.swmansion.com/scarb/download#install-via-asdf

I installed asdf and it solved the problem I described. Still think it's silly I needed another tool here or even to need to switch scarb versions at all. Surely foundry won't have this issue?

For example, say I'm working on two different projects in parallel, one that uses 0.5.1 and another that uses 0.6.0. I have to switch scarb versions before I run the scarb command for each repo, even if they're in different terminals. I am often working across repositories because in addition to Ekubo's AMM contracts, I also am working on a set of governance contracts.

**Comment by maciektr** [comment]
> I have to switch scarb versions before I run the scarb command for each repo

Not really! You can create '.tool-versions' file in the root of your repository with 'scarb <version>' line inside. The ASDF will handle switching automatically. That's the beauty of it! 

Please take a look https://asdf-vm.com/manage/configuration.html#tool-versions

**Comment by moodysalem** [comment]
> > I have to switch scarb versions before I run the scarb command for each repo
> 
> Not really! You can create '.tool-versions' file in the root of your repository with 'scarb ' line inside. The ASDF will handle switching automatically. That's the beauty of it!
> 
> Please take a look https://asdf-vm.com/manage/configuration.html#tool-versions

Got it. Maybe I'm just being stubborn, this still feels different unnecessarily imo, but I don't have any other unsolved issues with this approach with the tools you've suggested. Thanks!

**Comment by mkaput** [comment]
I think our divergence here simply comes from the fact that when architecting Cairo & Starknet developer experience about 1,5 years ago, we decided not to follow Ethereum patterns and instead go for the way how Rust is done. I've also mixed a pinch of influences from Elixir world. Reason? Users' perception of the tooling of Rust/Elixir is generally more positive than what of Ethereum has, and our job is to make the best things possible and not to reinvent wheels too much.

Cairo/Starknet tools very much map to what we have in Rust:

1. Cairo compiler → rustc = a raw compiler, reference implementation of the language, used practically only by higher layers
2. Scarb → Cargo = a user-focused toolchain which is installable, like the go CLI
3. asdf w/ scarb plugin → rustup = an application for installing & managing toolchains

But there are differences, as outlined in my previous responses here. For example, in Rust the actual toolchain is called a "toolchain" and Cargo is the main binary within it. Or, that the raw compiler is created by a different organization than the toolchain and others, but we have appropriate agreements on this. These are nuances, IMO.

But the unchanged thing is, that we have a separate layer for installing & managing the toolchain; which in Ethereum world is convoluted into the toolchain layer (Foundry is the primary example). Is our approach an extra step? Perhaps. Does it decouple concerns? Definitely yes. Which one is better? Time will tell, I think.

---

Pushing asdf is actually one of this of little spices from Elixir world, where asdf is de facto the "official" version manager. It gives us a lot of stuff for free (check out how much code rustup has), and also makes our tools follow some popular "standards".
